File: YEAR 2/SEM 1/Advanced Programming Methods/LAB 4/src/Model/Statement/ReadFileStatement.java
package Model.Statement;

import Model.Expression.ExpressionException;
import Model.Expression.IExpression;
import Model.State.ProgramState;
import Util.Dictionary.MyIDictionary;
import Util.Tuple.Tuple;

import java.io.BufferedReader;
import java.io.IOException;

public class ReadFileStatement implements IStatement {
    private String varName;
    private IExpression expFileId;

    public ReadFileStatement(IExpression expFileId,String varName) {
        this.expFileId = expFileId;
        this.varName = varName;
    }
    @Override
    public ProgramState execute(ProgramState programState) throws ExpressionException, StatementException {
        MyIDictionary<String,Integer> symbolTable = programState.getSymbolTable();

        int uniqueId = expFileId.evaluate(symbolTable,programState.getHeap());
        Tuple<String, BufferedReader> tuple = programState.getFileTable().get(uniqueId);

        if( tuple == null){
            throw new StatementException("Could not open file!");
        }

        BufferedReader bufferedReader = tuple.getSecond();

        String line;
        try {
            line = bufferedReader.readLine();
        }catch(IOException ex) {
            throw new StatementException("An error occurred while reading the file!");
        }

        int value = (line == null) ? 0 : Integer.parseInt(line);

        symbolTable.put(this.varName,value);

        return null;
    }

    @Override
    public String toString() {
        return "ReadFileStatement(" + this.varName + "," + this.expFileId + ")";
    }
}

File: YEAR 2/SEM 1/Advanced Programming Methods/ToyLanguage/ToyLanguage/Model/Statement/ReadFileStatement.cs
using System;
using System.IO;
using ToyLanguage.Model.Expression;
using ToyLanguage.Model.State;
using ToyLanguage.Util.Dictionary;

namespace ToyLanguage.Model.Statement
{
    class ReadFileStatement : IStatement
    {
        private readonly string varName;
        private readonly IExpression expFileId;
        public ReadFileStatement(IExpression expFileId,string varName)
        {
            this.expFileId = expFileId;
            this.varName = varName;
        }

        public ProgramState Execute(ProgramState programState)
        {
            MyIDictionary<string, int> symbolTable = programState.SymbolTable;

            int fileDescriptor = expFileId.Evaluate(symbolTable);

            if(!programState.FileTable.TryGetValue(fileDescriptor, out Tuple<string, TextReader> value))
            {
                throw new StatementException("Invalid file id provided!");
            }

            TextReader reader = value.Item2;
            string line;

            try
            {
                line = reader.ReadLine();
            }catch(Exception)
            {
                throw new StatementException("Error while reading from file!");
            }

            int intValue = 0;

            if (line != null && !int.TryParse(line, out intValue))
            {
                throw new StatementException("File contains non-numeric values!");
            }

            if (symbolTable.ContainsKey(varName))
                symbolTable.Update(varName, intValue);
            else
                symbolTable.Add(varName, intValue);

            return programState;
        }

        public override string ToString()
        {
            return "ReadFileStatement(" + expFileId + "," + varName + ")";
        }
    }
}
